### PR TITLE
stdenv: fix missing dependencies in __sandboxProfile and __impureHostDeps

### LINKED ISSUE #237458: propagatedSandboxProfile does not appear to propagate
### Describe the bug

While working on #234122, I attempted to use `propagatedSandboxProfile` in the qt5.qtbase derivation. Unfortunately, it did not appear to propagate to qtpass or other Qt-using packages that would require the profile. They still experienced the error until I manually added sandbox profiles to the other packages.

### Steps To Reproduce
1. Add a `propagatedSandboxProfile` to a package for something it needs (e.g., qt5.qtbase).
2. Use that package as a `buildInput` or `nativeBuildInput` to another package (e.g., qtpass).
3. Build the other package.

### Expected behavior

The build should succeed (allowing whatever is permitted by the profile), but it fails (due to the lack of the sandbox profile).

### Screenshots

n/a

### Additional context

This requires the sandbox on Darwin to be set to `true` or `relaxed`.

### Metadata

```console
$ nix-shell -p nix-info --run "nix-info -m"
 - system: `"aarch64-darwin"`
 - host os: `Darwin 22.5.0, macOS 13.4`
 - multi-user?: `yes`
 - sandbox: `relaxed`
 - version: `nix-env (Nix) 2.11.1`
 - channels(root): `"nixpkgs"`
 - nixpkgs: `/etc/nix/inputs/nixpkgs`
```

### PR DESCRIPTION
**Pull request by corngood**
Fixes: 7f3ca3e21a22 (stdenv: Fix handling of dependencies and hooks)
Fixes: #237458

## Description of changes

These seem to have been broken ages ago.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [x] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) (or backporting [23.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md) and [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

**Review by philiptaron — APPROVED** [review]
I built `glib` (as a random heavily-depended-on package), `musescore`, `qtpass`, `qtkeychain`, and all their dependencies on `x86_64-linux` with this change.

Thanks for reducing one super irritating edge in Nixpkgs!

**Comment by corngood** [comment]
I would love to `nixpkgs-review` this, but the only darwin machine I have access to is the community builder, and it was likely going to run out of storage.  I did build a good chunk of things with sandboxing enabled, and didn't hit any problems.

**Comment by reckenrode** [comment]
Looks like there’s a merge conflict now.

**Comment by corngood** [comment]
Apologies, I accidentally pushed a rebase on master, which added a bunch of reviewers. :(

File: pkgs/stdenv/generic/make-derivation.nix [hunk 1]
```diff
@@ -317,25 +317,28 @@ else let
     ]
   ];
 
+  allDependencies = concatLists (concatLists dependencies);
+  allPropagatedDependencies = concatLists (concatLists propagatedDependencies);
```
[the remark this answers is not in the document]
Comment (doronbehar) [reply pkgs/stdenv/generic/make-derivation.nix]: > The only other uses of `dependencies` and `propogatedDependencies` are down below:
> 
> [#291456 (files)](https://github.com/NixOS/nixpkgs/pull/291456/files#diff-282a02cc3871874f16401347d8fadc90d59d7ab11f6a99eaa5173c3867e1a160R399-R411)
> 
> I really wish the GitHub PR UI let me comment on them to highlight them.

So you are approving the changes, but are you implying something needs to change in these lines? I'm not sure I understand.

[the remark this answers is not in the document]
Comment (philiptaron) [reply pkgs/stdenv/generic/make-derivation.nix]: No, I'm griping about the GitHub code review UI that doesn't let me leave informative comments for other reviewers, who may be seeking out how to understand this change. Apologies for a confusing message.

**Review by philiptaron — APPROVED** [review]

**Comment by doronbehar** [comment]
Let's let this cook in `staging`.

**Pull request merged**